packages/embark/src/cmd/cmd_controller.js

```diff
@@ -504,58 +504,73 @@ class EmbarkController {
         }
 
-        engine.startService("web3");
-        engine.startService("deployment");
-        engine.startService("codeGenerator");
-        engine.startService("codeRunner");
-        engine.startService("console");
+        engine.registerModuleGroup("coreComponents");
+        engine.registerModuleGroup("stackComponents");
 
+        engine.registerModuleGroup("blockchain");
+        engine.registerModuleGroup("compiler");
+        engine.registerModuleGroup("contracts");
+        engine.registerModuleGroup("pipeline");
+        engine.registerModuleGroup("filewatcher");
+        engine.registerModuleGroup("storage");
+        engine.registerModuleGroup("communication");
+        engine.registerModulePackage('embark-deploy-tracker', {plugins: engine.plugins});
+        engine.registerModulePackage('embark-plugin-cmd', {embarkConfigFile: engine.embarkConfig, embarkConfig: engine.config.embarkConfig, packageFile: 'package.json'});
+        callback();
+      },
+      callback => {
         if (isSecondaryProcess(engine)) {
           return callback();
         }
-        engine.startService("embarkListener");
-        engine.startService("processManager");
-        engine.startService("coreProcess");
-        engine.startService("serviceMonitor");
-        engine.startService("libraryManager");
-        engine.startService("pipeline");
-        engine.startService("storage");
-        engine.startService("cockpit");
-        engine.startService("pluginCommand");
 
-        engine.events.request('blockchain:ready', callback);
-      },
-      function ipcConnect(callback) {
-        return callback();
-      },
-      function deploy(callback) {
-        // Skip if we are connected to a websocket, the server will do it
-        if(isSecondaryProcess(engine)) {
-          return callback();
-        }
-        engine.config.reloadConfig();
-        engine.events.request('deploy:contracts', function (err) {
-          callback(err);
+        let plugin = engine.plugins.createPlugin('cmdcontrollerplugin', {});
+        plugin.registerActionForEvent("embark:engine:started", async (_params, cb) => {
+          await engine.events.request2("blockchain:node:start", engine.config.blockchainConfig, cb);
         });
-      },
-      function waitForWriteFinish(callback) {
-        // Skip if we are connected to a websocket, the server will do it
-        if(isSecondaryProcess(engine)) {
-          return callback();
-        }
-        engine.logger.info("Finished deploying".underline);
-        engine.events.once('outputDone', (err) => {
-          engine.logger.info(__("finished building").underline);
-          callback(err);
+        plugin.registerActionForEvent("embark:engine:started", async (_params, cb) => {
+          await engine.events.request2("storage:node:start", engine.config.storageConfig, cb);
+        });
+        plugin.registerActionForEvent("embark:engine:started", async (_params, cb) => {
```

Comment:
Can put all three requests in the same `plugin.registerActionForEvent` and use `Promise.all` if you want parallelism.
eg: https://github.com/embark-framework/embark/pull/1782/files#diff-fe24036896987f2194f0d212d1241eb6L188-R201

Reply:
Ah thanks! I didn't expect the `cb` in `request2()` to be optional, which is why I didn't go for Promise.all().